File: Robot_Control_Course/Chapter14/Lesson3/Chapter14_Lesson3.m

function qdd_star = contactConsistentAcc(M, Jc, Jc_dot_qdot, qdd_des)
%CONTACTCONSISTENTACC Compute contact-consistent acceleration
%   M            : (n,n) mass matrix
%   Jc           : (nc,n) contact Jacobian
%   Jc_dot_qdot  : (nc,1) = Jc_dot(q,qdot) * qdot
%   qdd_des      : (n,1) unconstrained acceleration

Minv = inv(M);
Lambda_c = inv(Jc * Minv * Jc.');      % operational contact inertia
Jc_dyn_pinv = Minv * Jc.' * Lambda_c;  % M-consistent pseudo-inverse
Nc = eye(size(M,1)) - Jc_dyn_pinv * Jc;

qdd_star = Nc * qdd_des - Jc_dyn_pinv * Jc_dot_qdot;
end

File: Tutorials/Codes/Robot_Control_Course/Chapter14/Lesson3/Chapter14_Lesson3.cpp

#include <Eigen/Dense>

using Eigen::MatrixXd;
using Eigen::VectorXd;

MatrixXd dynConsistentPinv(const MatrixXd& M, const MatrixXd& Jc)
{
    // Jc_dyn_pinv = M^{-1} Jc^T (Jc M^{-1} Jc^T)^{-1}
    MatrixXd Minv = M.inverse();
    MatrixXd temp = Jc * Minv * Jc.transpose();
    MatrixXd Lambda_c = temp.inverse();
    MatrixXd Jc_dyn_pinv = Minv * Jc.transpose() * Lambda_c;
    return Jc_dyn_pinv;
}

void contactProjector(const MatrixXd& M,
                      const MatrixXd& Jc,
                      MatrixXd& Nc,
                      MatrixXd& Jc_dyn_pinv)
{
    int n = static_cast<int>(M.rows());
    Jc_dyn_pinv = dynConsistentPinv(M, Jc);
    Nc = MatrixXd::Identity(n, n) - Jc_dyn_pinv * Jc;
}

VectorXd contactConsistentAcc(const MatrixXd& M,
                              const MatrixXd& Jc,
                              const VectorXd& Jc_dot_qdot,
                              const VectorXd& qdd_des)
{
    MatrixXd Nc, Jc_dyn_pinv;
    contactProjector(M, Jc, Nc, Jc_dyn_pinv);
    // qdd_star = Nc qdd_des - Jc_dyn_pinv Jc_dot_qdot
    return Nc * qdd_des - Jc_dyn_pinv * Jc_dot_qdot;
}
